Converter for Uniswap V2 exchanges

roberts-pumpurs opened the pull request:
The said converter only performs price estimations via `estimateConvert` method - actual conversion (via `convert()`) is never done.

Commit:
build: explicitly add `solidity` compiler as a dependency

the version is pegged to the one used in the contracts

Commit:
build: introduced Uniswap interface contracts

Commit:
feat(converter): Initital ParsiqPancakeConverter implementation

It can only estimate token prices. Does not perform the actual convertation.

Commit:
docs(converter): added natspec comments to `ParsiqPancakeConverter`

Commit:
tests(parsiq-pancakeswap-conv): basic unnittests for making sure the pancakeswap comms are working

Commit:
feat(converter): using a new formula for calculating prices

Commit:
feat(converter): can use the prq pancakeswap converter both ways

By both ways: BUSD->PRQ and PRQ->BUSD

Commit:
tests(converter): make sure that converter works both ways

Commit:
refactor(converter): uniswap interface contracts moved to a subfolder

syakunin commented on contracts/ParsiqPancakeConverter.sol:
```diff
@@ -0,0 +1,97 @@
+ * Pancakeswap converter for estimating token prices.
+ */
+contract ParsiqPancakeConverter is IConverter {
+    IUniswapV2Pair public swapPair;
```
You should mark this as `immutable`

roberts-ivanovs replied:
Done

syakunin commented on contracts/ParsiqPancakeConverter.sol:
```diff
@@ -0,0 +1,97 @@
+ */
+contract ParsiqPancakeConverter is IConverter {
+    IUniswapV2Pair public swapPair;
+    IUniswapV2Router02 private _uniswapRouter;
```
`immutable`

roberts-ivanovs replied:
This could not be made as `immutable` because methods get called on the `_uniswapRouter` variable inside the constructor.

syakunin replied:
You can call on `uniswapRouter` parameter

roberts-ivanovs replied:
Thanks for the tip, updated the code :+1:

syakunin commented:
Are we creating converter per pair? I thought we will have universal converter with multiple pairs.

syakunin commented on contracts/ParsiqPancakeConverter.sol:
```diff
@@ -0,0 +1,97 @@
+        uint256 amountInSourceTokens,
+        IERC20 target
+    ) external pure override returns (uint256) {
+        require(address(source) == address(target), Errors.DC_UNSUPPORTED_PAIR);
```
No. If `source == target` you just return `amountInSourceTokens`

syakunin commented on contracts/ParsiqPancakeConverter.sol:
```diff
@@ -0,0 +1,97 @@
+        uint256 amountInSourceTokens,
+        IERC20 target
+    ) external view override returns (uint256) {
+        address pairToken0 = swapPair.token0();
```
Please add check for `source == target` and return `amountInSourceTokens` if it is true

roberts-ivanovs replied:
Alright, added it. Although this also means that now you can pass tokens to the converter that initially were not registered and no exception will be raised.

Commit:
refacot(converter): made the swap variable as immutable

Commit:
tests: make sure that you can estimate covnersions between unregistered tokens

roberts-ivanovs commented:
> Are we creating converter per pair? I thought we will have universal converter with multiple pairs.

Currently yes. This means that new enterprises will need to deploy their own instances of the converters if they wish to support their own tokens.

Commit:
refactor(converter): `_uniswapRouter` set as `immutable`

syakunin commented on package.json:
```diff
@@ -47,7 +47,8 @@
     "typechain": "^5.1.2",
-    "typescript": "^4.4.3"
+    "typescript": "^4.4.3",
+    "solc": "0.8.4"
```
Why this needs to be included?

roberts-ivanovs replied:
I noticed that there was no Solidity compiler installed locally for the project. Depending on your development environment setup, it can be configured to either use a globally installed Solidity compiler or a locally installed one for different tasks like linting or compiling. 
Specifying the exact Solidity compiler package allows for easier reproducible builds as it guarantees that there are no clashes with the necessary Solc and the systems installed Solc.

syakunin replied:
Hardhat installs correct solidity compiler automatically.

Commit:
refactor(@uniswap): using an npm package as a dependency

Commit:
feat(converter): initial implementation for the uniswap interface converter

Commit:
feat(converter): fully working `convert` method implementation

Commit:
refactor(naming): renamed `ParsiqConverter` to `SimpleConverter`

roberts-ivanovs commented:
Added a method to perform token conversions via the registered exchange.

Commit:
refactor(converter): renamed the `SimpleConverter` converter to `SimpleUniswapV2Converter`

Commit:
tests(converter): template for Enterprise integration tests

syakunin committed:
Fixed coverter test

syakunin committed:
Tests for converter

Commit:
revert(bf4c99c): the solidity compiler gets auto-downloaded by hardhat

it's up to the end-programmer to configure his dev envirionment to use the downloaded one instead of local node_modules one
